File: plugins/robots/robotsGenerator/nxtOSEK/generators/simpleElements/initialBlockGenerator.cpp
#include "initialBlockGenerator.h"
#include "../../nxtOSEKRobotGenerator.h"

using namespace robots::generator;

InitialBlockGenerator::InitialBlockGenerator()
{
}

QList<SmartLine> InitialBlockGenerator::convertElementIntoDirectCommand(NxtOSEKRobotGenerator *nxtGen
		, qReal::Id const elementId, qReal::Id const logicElementId)
{
	QList<SmartLine> result;
	QList<SmartLine> initCode;
	QList<SmartLine> terminateCode;
	int const numberOfPorts = 4;
	for (int i = 1; i <= numberOfPorts; i++) {
		QString const curPort = "port_" + QString::number(i);
		QByteArray portValue = nxtGen->api()->stringProperty(logicElementId, curPort).toUtf8();
		QString const initEcrobotColorSensorPortS = "ecrobot_init_nxtcolorsensor(NXT_PORT_S";
		if (portValue == "Ультразвуковой сенсор") {
			initCode.append(SmartLine("ecrobot_init_sonar_sensor(NXT_PORT_S" + QString::number(i) + ");"
					, elementId));
			terminateCode.append(SmartLine("ecrobot_term_sonar_sensor(NXT_PORT_S" + QString::number(i) + ");"
					, elementId));
		} else if (portValue == "Сенсор цвета (все цвета)") {
			initCode.append(SmartLine(initEcrobotColorSensorPortS + QString::number(i)
				+ ", NXT_LIGHTSENSOR_WHITE);", elementId));
			terminateCode.append(SmartLine("ecrobot_term_nxtcolorsensor(NXT_PORT_S"
			+ QString::number(i) + ");", elementId));
		} else if (portValue == "Сенсор цвета (красный)") {
			initCode.append(SmartLine(initEcrobotColorSensorPortS + QString::number(i)
				+ ",  NXT_LIGHTSENSOR_RED);", elementId));
			terminateCode.append(SmartLine("ecrobot_term_nxtcolorsensor(NXT_PORT_S"
			+ QString::number(i) + ");", elementId));
		} else if (portValue == "Сенсор цвета (зеленый)") {
			initCode.append(SmartLine(initEcrobotColorSensorPortS + QString::number(i)
				+ ", NXT_LIGHTSENSOR_GREEN);", elementId));
			terminateCode.append(SmartLine("ecrobot_term_nxtcolorsensor(NXT_PORT_S"
			+ QString::number(i) + ");", elementId));
		} else if (portValue == "Сенсор цвета (синий)") {
			initCode.append(SmartLine(initEcrobotColorSensorPortS + QString::number(i)
				+ ", NXT_LIGHTSENSOR_BLUE);", elementId));
			terminateCode.append(SmartLine("ecrobot_term_nxtcolorsensor(NXT_PORT_S"
			+ QString::number(i) + ");", elementId));
		} else if (portValue == "Сенсор цвета (пассивный)") {
			initCode.append(SmartLine(initEcrobotColorSensorPortS + QString::number(i)
				+ ", NXT_COLORSENSOR);", elementId));
			terminateCode.append(SmartLine("ecrobot_term_nxtcolorsensor(NXT_PORT_S"
			+ QString::number(i) + ");", elementId));
		}
	}
	nxtGen->initCode().append(initCode);
	nxtGen->terminateCode().append(terminateCode);

	return result;
}

File: plugins/visualInterpreter/examples/robotsCodeGeneration/reactionsStorage/InitialBlockGenerator.py
template = "#include \"kernel.h\"\n#include \"ecrobot_interface.h\"\n@@BALANCER@@\n@@VARIABLES@@\n\nvoid ecrobot_device_initialize(void)\n{\n@@INITHOOKS@@\n}\n\nvoid ecrobot_device_terminate(void)\n{\n@@TERMINATEHOOKS@@\n}\n\n/* nxtOSEK hook to be invoked from an ISR in category 2 */\nvoid user_1ms_isr_type2(void){ /* do nothing */ }\n\n@@CODE@@"

task_template = "TASK(OSEK_Task_Number_0)\n{\n@@CODE@@\n}"
template = template.replace("@@CODE@@", task_template)

number_of_ports = 4
port_values = [initBlock.port_1, initBlock.port_2, initBlock.port_3, initBlock.port_4]

for i in xrange(number_of_ports):
  init_ecrobot_color_sensor_port_s = "ecrobot_init_nxtcolorsensor(NXT_PORT_S"
  if port_values[i] == "Ультразвуковой сенсор":
    init_code.append("ecrobot_init_sonar_sensor(NXT_PORT_S" + str(i + 1) + ");\n")
    terminate_code.append("ecrobot_term_sonar_sensor(NXT_PORT_S" + str(i + 1) + ");\n")
    port_values[i] = "ecrobot_get_sonar_sensor(NXT_PORT_S"
  elif port_values[i] == "Сенсор цвета (все цвета)":
    init_code.append(init_ecrobot_color_sensor_port_s + str(i + 1) +", NXT_LIGHTSENSOR_WHITE);\n")
    terminate_code.append("ecrobot_term_nxtcolorsensor(NXT_PORT_S" + str(i + 1) + ");\n")
    port_values[i] = "ecrobot_get_light_sensor(NXT_PORT_S"
  elif port_values[i] == "Сенсор цвета (красный)":
    init_code.append(init_ecrobot_color_sensor_port_s + str(i + 1) + ", NXT_LIGHTSENSOR_RED);\n")
    terminate_code.append("ecrobot_term_nxtcolorsensor(NXT_PORT_S" + str(i + 1) + ");\n")
    port_values[i] = "ecrobot_get_light_sensor(NXT_PORT_S"
  elif port_values[i] == "Сенсор цвета (зеленый)":
    init_code.append(init_ecrobot_color_sensor_port_s + str(i + 1) + ", NXT_LIGHTSENSOR_GREEN);\n")
    terminate_code.append("ecrobot_term_nxtcolorsensor(NXT_PORT_S" + str(i + 1) + ");\n")
    port_values[i] = "ecrobot_get_light_sensor(NXT_PORT_S"
  elif port_values[i] == "Сенсор цвета (синий)":
    init_code.append(init_ecrobot_color_sensor_port_s + str(i + 1) + ", NXT_LIGHTSENSOR_BLUE);\n")
    terminate_code.append("ecrobot_term_nxtcolorsensor(NXT_PORT_S" + str(i + 1) + ");\n")
    port_values[i] = "ecrobot_get_light_sensor(NXT_PORT_S"
  elif port_values[i] == "Сенсор цвета (пассивный)":
    init_code.append(init_ecrobot_color_sensor_port_s + str(i + 1) + ", NXT_COLORSENSOR);\n")
    terminate_code.append("ecrobot_term_nxtcolorsensor(NXT_PORT_S" + str(i + 1) + ");\n")
    port_values[i] = "ecrobot_get_light_sensor(NXT_PORT_S"
  else:
    port_values[i] = "ecrobot_get_touch_sensor(NXT_PORT_S"

initBlock.id = max_used_id

cur_node_is_processed = True
